sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -453,45 +453,89 @@ case class RangeExec(range: org.apache.spark.sql.catalyst.plans.logical.Range)
+         |  ${consume(ctx, Seq(ev))}
+         |}
+         |$nextIndex = $batchEnd;
+         |$numOutput.add($localEnd);
+         |$inputMetrics.incRecordsRead($localEnd);
+       """.stripMargin
+    }
+
+    val loopCondition = if (parent.needStopCheck) {
+      "!shouldStop()"
     } else {
-      "// shouldStop check is eliminated"
+      "true"
     }
+
+    // An overview of the Range processing.
+    //
+    // For each partition, the Range task needs to produce records from partition start(inclusive)
+    // to end(exclusive). For better performance, we separate the partition range into batches, and
+    // use 2 loops to produce data. The outer while loop is used to iterate batches, and the inner
+    // for loop is used to iterate records inside a batch.
+    //
+    // `nextIndex` tracks the index of the next record that is going to be consumed, initialized
+    // with partition start. `batchEnd` tracks the end index of the current batch, initialized
+    // with `nextIndex`. In the outer loop, we first check if `batchEnd - nextIndex` is non-zero.
+    // Note that it can be negative, because range step can be negative. If `batchEnd - nextIndex`
+    // is non-zero, we enter the inner loop. Otherwise, we update `batchEnd` to process the next
+    // batch. If `batchEnd` reaches partition end, exit the outer loop. Since `batchEnd` is
+    // initialized with `nextIndex`, the first iteration of outer loop will not enter the inner
+    // loop but just update the `batchEnd`.
+    //
+    // The inner loop iterates from 0 to `localEnd`, which is calculated by
+    // `(batchEnd - nextIndex) / step`. Since `batchEnd` is increased by `nextBatchTodo * step` in
+    // the outer loop, and initialized with `nextIndex`, so `batchEnd - nextIndex` is always
+    // divisible by `step`. The `nextIndex` is increased by `step` during each iteration, and ends
+    // up being equal to `batchEnd` when the inner loop finishes.
+    //
+    // The inner loop can be interrupted, if the query has produced at least one result row, so that
+    // we don't buffer many result rows and waste memory. It's ok to interrupt the inner loop,
+    // because `nextIndex` is updated per loop iteration and remembers how far we have processed.
+
     s"""
       | // initialize Range
       | if (!$initTerm) {
       |   $initTerm = true;
       |   $initRangeFuncName(partitionIndex);
       | }
       |
-      | while (true) {
-      |   long $range = $batchEnd - $number;
+      | while ($loopCondition) {
+      |   long $range = $batchEnd - $nextIndex;
       |   if ($range != 0L) {
       |     int $localEnd = (int)($range / ${step}L);
-      |     for (int $localIdx = 0; $localIdx < $localEnd; $localIdx++) {
-      |       long $value = ((long)$localIdx * ${step}L) + $number;
-      |       ${consume(ctx, Seq(ev))}
-      |       $shouldStop
+      |     $processingLoop
+      |   } else {
+      |     long $nextBatchTodo;
```

Comment:
why did you move these lines in the else?

Reply:
Now we don't do `return` when we need to stop the loop. Move these lines to else, so that we won't hit this code path when loop is interrupted.

Reply:
I see, but in this way we are looping 2 more times in the outer loop, because we either go in the if or in the else while previously we were doing both on the same iteration IIUC. I don't think it is a big issue but it may introduce a (very small probably) overhead compared to the previous case.

Since if IIUC in the first iteration we just go to the else branch now, since `batchEnd` is inited to `nextIndex`, do you think it is feasible to move this block before the inner loop? So we would solve both issues, right?

Reply:
good idea!